### PR TITLE
Bci gui rewrite

### PR DESCRIPTION
**Pull request by Carsonthemonkey**
# Overview

Rewrite BCIGui code to be simpler and more maintainable

## Ticket

[GUI Refactor](https://www.pivotaltracker.com/story/show/188031090)
[Intertask Action](https://www.pivotaltracker.com/story/show/187994028)

## Contributions

- Rewrite base GUI class
- Rewrite Experiment registry to extend base gui class
- Create intertask action with new ui code

## Test

- Manual testing (may want to look into Qt Test in the future for automated GUI testing)

## Documentation

- Inline documentation written for GUI methods

## Changelog

- Changelog updated

### CODE REVIEW

File: bcipy/gui/bciui.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+    def list(self):
+        return [widget.data for widget in self.widgets]
+
+    def list_property(self, prop: str):
+        return [widget.data[prop] for widget in self.widgets]
+
+
+# --- Experiment registry code ---
+from bcipy.helpers.load import load_fields, load_experiments
```
Comment (tab-cmd) [inline bcipy/gui/bciui.py]: These imports should be at the top of the file (PEP). They can still be separated with a comment as you have done here if that's helpful.

File: bcipy/gui/bciui.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
+import os
```
Comment (tab-cmd) [inline bcipy/gui/bciui.py:1]: `gui.py` or `main.py` are possible filenames. This one works though. We should deprecate older files in this PR. If the filename changes, we'll also need to update the Makefile!

File: bcipy/gui/bcipy_stylesheet.css [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/* This stylesheet uses the QSS syntax, but is named as a CSS file to take advantage of IDE CSS tooling */
```
Comment (tab-cmd) [inline bcipy/gui/bcipy_stylesheet.css:1]: I really like this change!

File: bcipy/gui/bciui.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    QMessageBox,
+)
+from typing import Optional, List
+
+
+class BCIUI(QWidget):
+    contents: QVBoxLayout
+    center_content_vertically: bool = False
+
+    def __init__(self, title="BCIUI", default_width=500, default_height=600):
```
Comment (tab-cmd) [inline bcipy/gui/bciui.py]: types/ returns needed

File: bcipy/gui/bciui.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+        super().__init__()
+        self.resize(default_width, default_height)
+        self.setWindowTitle(title)
+        self.contents = QVBoxLayout()
+        self.setLayout(self.contents)
+
+    def app(self):
```
Comment (tab-cmd) [inline bcipy/gui/bciui.py]: Needed?

File: bcipy/gui/bciui.py (hunk 1 above)
Comment (tab-cmd) [reply bcipy/gui/bciui.py]: This could also be a separate file.

File: bcipy/gui/intertask_action.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+from PyQt6.QtWidgets import (
+    QApplication,
+    QLabel,
+    QHBoxLayout,
+    QPushButton,
+    QProgressBar,
+)
+
+
+class IntertaskAction(BCIUI):
```
Comment (tab-cmd) [inline bcipy/gui/intertask_action.py]: This should likely be InterTask GUI, the Action itself will need to inherit Task.

File: bcipy/gui/bciui.py (hunk 1 above)
Comment (Carsonthemonkey) [reply bcipy/gui/bciui.py]: Planning to split this into a seperate file before merging. Should I delete the original `ExperimentRegistry` on this branch before merging?

File: bcipy/gui/bciui.py (hunk 5 above)
Comment (Carsonthemonkey) [reply bcipy/gui/bciui.py]: Maybe I should make this an abstract method

Comment (Carsonthemonkey) [reply bcipy/gui/bciui.py]: Seems like I can't inherit from both `QWidget` and `ABC` so leaving this in for now

File: bcipy/gui/bciui.py (hunk 1 above)
Comment (tab-cmd) [reply bcipy/gui/bciui.py]: That sounds good to me, let's deprecate it!

File: bcipy/gui/intertask_gui.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+from PyQt6.QtWidgets import (
+    QApplication,
+    QLabel,
+    QHBoxLayout,
+    QPushButton,
+    QProgressBar,
+)
+
+
+class IntertaskGUI(BCIUI):
```
Comment (tab-cmd) [inline bcipy/gui/intertask_gui.py:15]: Wouldn't this need to be a task to use this in orchestration? Or is this going to be a feature that is either on or off? If that's the case, we may want to add a boolean to protocol/experiment registry to configure this.

**Pull request merged**